src/main/java/com/google/step/snippet/servlets/UserSettingServlet.java

```diff
@@ -24,11 +29,38 @@
 /** Servelet that renders user setting page */
 @WebServlet("/user")
 public class UserSettingServlet extends HttpServlet {
+  private static final String WEBSITE_PARAMETER = "website";
+  private static final String LANGUAGE_PARAMETER = "language";
+  private static final String ID_PARAMETER = "id";
+  private static final String USER_PARAMETER = "UserInfo";
+  private String referer = "";
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response)
       throws IOException, ServletException {
     // Forward the request to the template (which is a servlet itself).
     request.getRequestDispatcher("WEB-INF/templates/user_dashboard.jsp").forward(request, response);
+    referer = request.getHeader("referer");
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/home");
+      return;
+    }
+
+    String website = request.getParameter(WEBSITE_PARAMETER);
+    String language = request.getParameter(LANGUAGE_PARAMETER);
+    String id = userService.getCurrentUser().getUserId();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity entity = new Entity(USER_PARAMETER, id);
```

Comment:
Here's the proposed code modification I mentioned yesterday to only create a new user entity if one doesn't exist yet

```
String uid = userService.getCurrentUser().getUserId();
Query.FilterPredicate filterUser = new Query.FilterPredicate(UID, FilterOperator.EQUAL, uid);
Query queryUser = new Query(USER).setFilter(filterUser);
Entity userEntity = datastore.prepare(queryUser).asSingleEntity();

if (userEntity == null) {
      userEntity = new Entity(USER);
      userEntity.setProperty(UID, uid);
      userEntity.setProperty(EMAIL, userService.getCurrentUser().getEmail());
 }
...

entity.setProperty(WEBSITE_PARAMETER, website);
entity.setProperty(LANGUAGE_PARAMETER, language);
datastore.put(entity);
```